src/workspace-data/data-table/wds/RecordDeleter.js

```diff
@@ -0,0 +1,98 @@
+import _ from 'lodash/fp';
+import { useState } from 'react';
+import { b, div, h } from 'react-hyperscript-helpers';
+import { absoluteSpinnerOverlay, DeleteConfirmationModal } from 'src/components/common';
+import { Ajax } from 'src/libs/ajax';
+import colors from 'src/libs/colors';
+import { reportError } from 'src/libs/error';
+import * as Utils from 'src/libs/utils';
+
+export const RecordDeleter = ({ onDismiss, onSuccess, dataProvider, collectionId, selectedRecords, runningSubmissionsCount }) => {
+  const [additionalDeletions, setAdditionalDeletions] = useState([]);
+  const [deleting, setDeleting] = useState(false);
+
+  const selectedKeys = _.keys(selectedRecords);
+
+  const doDelete = async () => {
+    const recordsToDelete = _.flow(
+      _.map(({ name: entityName, entityType }) => ({ entityName, entityType })),
+      (records) => _.concat(additionalDeletions, records)
+    )(selectedRecords);
+
+    const recordTypes = _.uniq(_.map(({ entityType }) => entityType, selectedRecords));
+    if (recordTypes.length > 1) {
+      await reportError('Something went wrong; more than one recordType is represented in the selection. This should not happen.');
+    }
+    const recordType = recordTypes[0];
+    setDeleting(true);
+
+    try {
+      await Ajax().WorkspaceData.deleteRecords(dataProvider.proxyUrl, collectionId, recordType, {
+        record_ids: recordsToDelete,
+      });
+      onSuccess();
+    } catch (error) {
+      switch (error.status) {
+        case 409:
+          setAdditionalDeletions(
+            _.filter(
+              (errorEntity) =>
+                !_.some(
+                  (selectedEntity) => selectedEntity.entityType === errorEntity.entityType && selectedEntity.entityName === errorEntity.entityName,
+                  recordsToDelete
+                ),
+              await error.json()
+            )
+          );
+          setDeleting(false);
+          break;
+        default:
+          await reportError('Error deleting data entries', error);
+          onDismiss();
```

Comment:
Sonarcloud suggested removing the switch case: _"switch" statements should have at least 3 "case" clauses_
Use the new function `filterAdditionalDeletions` added above in your implementation

```suggestion
      if (error.status != 409){
         await reportError('Error deleting data entries', error);
         return onDismiss();
      }
          
      // Handle 409 error by filtering additional deletions that need to be deleted first
      setAdditionalDeletions(await filterAdditionalDeletions(error, recordsToDelete));
      setDeleting(false);
```